### ISSUE TITLE
Over sample in SGLang

### ISSUE THREAD

**Issue by zhaochenyang20** [zhaochenyang20/Awesome-ML-SYS-Tutorial#193]
## 检查

```python
            loss_mask=original_req.loss_mask,
            prompt_loss_mask=original_req.prompt_loss_mask,
            response_loss_mask=padding_response_loss_mask,
```

这几行就有问题，`loss_mask` 和 `prompt_loss_mask, response_loss_mask` 关系是啥？

`prompt_loss_mask` 应该是全 0，然后 `padding_response_loss_mask` 应该也是全 0。

```python
            prompt_attention_mask=original_req.prompt_attention_mask,
            response_attention_mask=padding_response_attention_mask,
```

`attention_mask` 而言，`prompt_attention_mask` 应该是全 1，而 `response_attention_mask` 是全 0，这样：

prompt 只算了 forward，没算 loss；而 response 不算 forward，也不算 loss。

**我不确定上面描述的是否符合预期。**

## 新的优化

如果上面的讨论正确的话，我想为了节省 forward 的消耗，做进一步的优化：

`prompt_attention_mask` 设为 `[1]`，意味着整个 prompt 只有 1 个 token 算 forward；然后对应 prompt 的 loss mask 是 `[0]`，意味着这唯一的 token 不算 loss。

我认为这样有两个好处：

1. 相比于之前完整的 `prompt_attention_mask`，被 padding 的 reqs 计算 forward 的 token 更少了；
2. 保留了一个 token，而不是一个 token 都不保留；避免在 `micro_batch_size_per_gpu=1` 的时候，整个 req 被 drop 掉，这个 gpu 上遇到一个空的 batch，直接报错；

我之前遇到的奇怪事情是，`micro_batch_size_per_gpu=1` 时，填为 padding 的 reqs 反而增加了训练时间，我后来想明白，这是因为在我那个版本的实现里面，`prompt_attention_mask` 设为了 `[1] * (prompt_max_length + response_max_length)`，虽然没算 loss，但是直接就让 forward 的 token 多算了很多。此外，如果我直接把 `prompt_attention_mask` 设为 `[]` 的话，会在 `micro_batch_size_per_gpu=1` 时整个 request 都被 drop，导致 batch 实际为空，从而训练失败。

当然，在开源的情况下，我们很少会设置 `micro_batch_size_per_gpu=1`，所以从未遇到过这个问题。

**Comment by zhaochenyang20** [comment]
如果模型生成了一个 tool 调用的请求，这个请求写错了，导致 tool calling 失败，这个时候这个 request 应该算作结束，而不是被 catch 住然后 padding 起来。这个 request 应该在训练阶段被使用，然后教会模型“不该这么 call tool”，而不是被我们 padding 起来，不对模型产生任何影响。

**Comment by PopSoda2002** [comment]
check `prompt_loss_mask` is used or not

**Comment by PopSoda2002** [comment]
what if `tool_call` failed?

**Comment by PopSoda2002** [comment]
```python
batch = TensorDict(
            {
                "prompts": idx,
                "responses": response,
                "input_ids": seq,  # here input_ids become the whole sentences
                "attention_mask": attention_mask,
                "position_ids": position_ids,
            },
            batch_size=batch_size,
        )
```
```python
# Construct the batch data
        batch = TensorDict(
            {
                "prompts": prompt_ids,
                "responses": response_ids,
                "response_mask": response_loss_mask,
                "input_ids": input_ids,  # here input_ids become the whole sentences
                "attention_mask": attention_mask,
                "position_ids": position_ids,
            },
            batch_size=len(sorted_output_req_list),
        )
```
其实我们只会返回这五-六个字段 很多字段都是没被使用
AgentLoop 中返回的batch 其实也是返回这些字段:
```python
# NOTE: consistent with batch version of generate_sequences in vllm_rollout_spmd.py
# prompt_ids: left padded with zeros (e.g., [0,0,0,0,1,2,3,4])
# response_ids: right padded with zeros (e.g., [5,6,7,8,0,0,0,0])
# input_ids: concatenation of prompt + response
# Mask:
# For example, if the prompt is [1,2,3,4] and the response is [5,6,7,(tool start)8,9(tool end),10,11,12]
# - prompt_attention_mask: 0s for padding, 1s for tokens
#   e.g., [0,0,0,0,1,1,1,1]
# - response_attention_mask: 0s for padding, 1s for tokens
#   e.g., [1,1,1,1,1,1,1,1,1,1,1,0,0,0,0]
# attention_mask: concatenation of prompt_attention_mask and response_attention_mask
#   e.g., [0,0,0,0,1,1,1,1(prompt),1,1,1,1,1,1,1,1,1,1,1,0,0,0,0(response)]
# - response_mask: 1s for LLM generated tokens, 0 for tool response/padding tokens
#   e.g., [1,1,1,1,1,1,1,(tool start),0,0(tool end),1,1,0,0,0,0]
# - position_ids: sequential positions for tokens, starting at 0
#   e.g., [0,0,0,0,0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,0,0,0,0]

batch = TensorDict(
            {
                "prompts": prompt_ids,  # [bsz, prompt_length]
                "responses": response_ids,  # [bsz, response_length]
                "response_mask": response_mask,  # [bsz, response_length]
                "input_ids": input_ids,  # [bsz, prompt_length + response_length]
                "attention_mask": attention_mask,  # [bsz, prompt_length + response_length]
                # position_ids: [bsz, 3, prompt_length + response_length] or [bsz, prompt_length + response_length]
                "position_ids": position_ids,
            },
            batch_size=len(inputs),
        )
```
